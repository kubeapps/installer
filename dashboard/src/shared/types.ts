import { IAuthState } from "../reducers/auth";
import { IServiceCatalogState } from "../reducers/catalog";
import { IConfigState } from "../reducers/config";
import { IFunctionState } from "../reducers/functions";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { hapi } from "./hapi/release";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file
export class ForbiddenError extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class NotFoundError extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class MissingChart extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class AppConflict extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
  };
  items: IChart[];
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IResource {
  apiVersion: string;
  kind: string;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret extends IResource {
  data: { [s: string]: string };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IFunction extends IResource {
  spec: {
    deps: string;
    function: string;
    handler: string;
    runtime: string;
    checksum: string;
  };
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: hapi.release.Release[];
  listOverview?: IAppOverview[];
  selected?: hapi.release.Release;
}

export interface IStoreState {
  catalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  config: IConfigState;
  repos: IAppRepositoryState;
  deployment: IDeployment;
  functions: IFunctionState;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
      {
        clusterName: string;
        creationTimestamp: string;
        deletionGracePeriodSeconds: string | null;
        deletionTimestamp: string | null;
        resourceVersion: string;
        selfLink: string;
      },
      {
        type: string;
        url: string;
        auth: {
          header: {
            secretKeyRef: {
              name: string;
              key: string;
            };
          };
        };
      },
      undefined
    > {}

export interface IAppRepositoryList
  extends IK8sList<
      IAppRepository,
      {
        continue: string;
        resourceVersion: string;
        selfLink: string;
      }
    > {}

export interface IFunctionList extends IK8sList<IFunction, {}> {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IKubelessConfigMap {
  metadata: {
    name: string;
  };
  data: {
    "runtime-images": string;
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
}
