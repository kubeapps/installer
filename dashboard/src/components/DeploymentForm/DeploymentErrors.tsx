import * as React from "react";
import {
  AppConflict,
  ForbiddenError,
  IRBACRole,
  MissingChart,
  NotFoundError,
  UnprocessableEntity,
} from "../../shared/types";
import { NotFoundErrorAlert, PermissionsErrorAlert, UnexpectedErrorAlert } from "../ErrorAlert";

interface IDeploymentErrorProps {
  kubeappsNamespace: string;
  namespace: string;
  releaseName: string;
  error: Error | undefined;
  chartName: string;
  repo: string;
  version: string;
}

class DeploymentErrors extends React.Component<IDeploymentErrorProps> {
  public render() {
    const { chartName, error, namespace, releaseName, repo, version } = this.props;
    switch (error && error.constructor) {
      case AppConflict:
        return (
          <NotFoundErrorAlert
            header={"The given release name already exists in the cluster. Choose a different one"}
          />
        );
      case ForbiddenError:
        const message = error ? error.message : "";
        let roles: IRBACRole[] = [];
        try {
          roles = JSON.parse(message);
        } catch (e) {
          // Cannot parse the error as a role array
          // return the default roles
          roles = this.requiredRBACRoles();
        }
        return (
          <PermissionsErrorAlert
            namespace={namespace}
            roles={roles}
            action={`deploy the application "${releaseName}"`}
          />
        );
      case MissingChart:
        return (
          <NotFoundErrorAlert header={`Chart ${chartName} (v${version}) not found in ${repo}`} />
        );
      case NotFoundError:
        return (
          <NotFoundErrorAlert resource={`Application "${releaseName}"`} namespace={namespace} />
        );
      case UnprocessableEntity:
        return (
          <UnexpectedErrorAlert
            text={error && error.message}
            raw={true}
            title="Failed installation"
          />
        );
      default:
        return <UnexpectedErrorAlert />;
    }
  }

  private requiredRBACRoles(): IRBACRole[] {
    return [
      {
        apiGroup: "kubeapps.com",
        namespace: this.props.kubeappsNamespace,
        resource: "apprepositories",
        verbs: ["get"],
      },
    ];
  }
}

export default DeploymentErrors;
