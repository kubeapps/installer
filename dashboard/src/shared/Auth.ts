import Axios, { AxiosRequestConfig } from "axios";

const AuthTokenKey = "kubeapps_auth_token";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string) {
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return ["base64url.bearer.authorization.k8s.io." + btoa(token).replace(/=*$/g, "")];
  }

  public static fetchOptions() {
    return { headers: { Authorization: `Bearer ${this.getAuthToken()}` } };
  }
}

// authenticatedAxiosInstance returns an axios instance with an interceptor
// configured to set the current auth token
function authenticatedAxiosInstance() {
  const a = Axios.create();
  a.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
  return a;
}

export const axios = authenticatedAxiosInstance();
