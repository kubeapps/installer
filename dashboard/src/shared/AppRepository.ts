import { axios } from "./Auth";
import Config from "./Config";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list() {
    const { data } = await axios.get<IAppRepositoryList>(await AppRepository.getResourceLink());
    return data;
  }

  public static async get(name: string) {
    const { data } = await axios.get(await AppRepository.getResourceLink(name));
    return data;
  }

  public static async update(name: string, newApp: IAppRepository) {
    const { data } = await axios.put(await AppRepository.getResourceLink(name), newApp);
    return data;
  }

  public static async delete(name: string) {
    const { data } = await axios.delete(await AppRepository.getResourceLink(name));
    return data;
  }

  public static async create(name: string, url: string, auth: any) {
    const { data } = await axios.post<IAppRepository>(await AppRepository.getResourceLink(), {
      apiVersion: "kubeapps.com/v1alpha1",
      kind: "AppRepository",
      metadata: {
        name,
      },
      spec: { auth, type: "helm", url },
    });
    return data;
  }

  private static APIBase: string = "/api/kube";
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static async getResourceLink(name?: string) {
    const { namespace } = await Config.getConfig();
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories${
      name ? `/${name}` : ""
    }`;
  }
}
