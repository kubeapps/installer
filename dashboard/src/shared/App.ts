import { AppRepository } from "./AppRepository";
import { axios } from "./Auth";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export class App {
  public static getResourceURL(namespace?: string, name?: string) {
    let url = "/api/tiller-deploy/v1";
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name, kubeappsNamespace);
    const auth = repo.spec.auth;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axios.post(endpoint, {
      auth,
      chartName: chartAttrs.name,
      releaseName,
      repoUrl: chartAttrs.repo.url,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name, kubeappsNamespace);
    const auth = repo.spec.auth;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axios.put(endpoint, {
      auth,
      chartName: chartAttrs.name,
      releaseName,
      repoUrl: chartAttrs.repo.url,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async delete(releaseName: string, namespace: string) {
    const { data } = await axios.delete(App.getResourceURL(namespace, releaseName));
    return data;
  }

  public static async listApps(namespace?: string) {
    const { data } = await axios.get<{ data: IAppOverview[] }>(App.getResourceURL(namespace));
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axios.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
