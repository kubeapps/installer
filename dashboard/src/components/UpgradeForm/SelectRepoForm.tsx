import * as React from "react";

import { ForbiddenError, IAppRepository, IRBACRole, MissingChart } from "../../shared/types";

import { NotFoundErrorAlert, PermissionsErrorAlert, UnexpectedErrorAlert } from "../ErrorAlert";

import "brace/mode/yaml";
import "brace/theme/xcode";

interface ISelectRepoFormProps {
  kubeappsNamespace: string;
  error: Error | undefined;
  repo: IAppRepository;
  repos: IAppRepository[];
  chartName: string;
  checkChart: (repo: string, chartName: string) => any;
}

interface ISelectRepoFormState {
  repo: string;
}

class SelectRepoForm extends React.Component<ISelectRepoFormProps, ISelectRepoFormState> {
  public state: ISelectRepoFormState = {
    repo:
      this.props.repo.metadata && this.props.repo.metadata.name
        ? this.props.repo.metadata.name
        : "",
  };

  public render() {
    if (!this.props.repos) {
      return <div> Loading </div>;
    }
    return (
      <div className="container margin-normal">
        <div className="col-8">{this.props.error && this.renderError()}</div>
        <div className="col-12">
          <h2>Select the source repository of {this.props.chartName}</h2>
        </div>
        <div className="col-8">
          <label htmlFor="chartRepoName">Chart Repository Name *</label>
          <select
            id="chartRepoName"
            onChange={this.handleChartRepoNameChange}
            value={this.state.repo}
            required={true}
          >
            {!this.state.repo && <option key="" value="" />}
            {this.props.repos.map(r => (
              <option key={r.metadata.name} value={r.metadata.name}>
                {r.metadata.name} ({this.getRepoURL(r.metadata.name)})
              </option>
            ))}
          </select>
        </div>
        <div>
          <p>
            {" "}
            * If the repository containing {this.props.chartName} is not in the list add it{" "}
            <a href="/config/repos"> here </a>.{" "}
          </p>
        </div>
      </div>
    );
  }

  public handleChartRepoNameChange = async (e: React.ChangeEvent<HTMLSelectElement>) => {
    this.props.checkChart(e.currentTarget.value, this.props.chartName);
    this.setState({ repo: e.currentTarget.value });
  };

  private getRepoURL = (name: string) => {
    let res = "";
    this.props.repos.forEach(r => {
      if (r.metadata.name === name && r.spec) {
        res = r.spec.url;
      }
    });
    return res;
  };

  private renderError() {
    if (this.props.error) {
      switch (this.props.error.constructor) {
        case MissingChart:
          return <NotFoundErrorAlert header={this.props.error.message} />;
        case ForbiddenError:
          return (
            <PermissionsErrorAlert
              namespace={this.props.kubeappsNamespace}
              roles={[this.requiredRBACRoles()]}
              action={`view App Repositories`}
            />
          );
      }
    }
    return <UnexpectedErrorAlert />;
  }

  private requiredRBACRoles(): IRBACRole {
    return {
      apiGroup: "kubeapps.com",
      namespace: this.props.kubeappsNamespace,
      resource: "apprepositories",
      verbs: ["get"],
    };
  }
}

export default SelectRepoForm;
