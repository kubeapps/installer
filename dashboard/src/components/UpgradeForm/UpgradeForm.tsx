import * as React from "react";
import AceEditor from "react-ace";
import { RouterAction } from "react-router-redux";

import { IServiceBindingWithSecret } from "../../shared/ServiceBinding";
import { IChartState, IChartVersion } from "../../shared/types";
import DeploymentBinding from "../DeploymentForm/DeploymentBinding";
import DeploymentErrors from "../DeploymentForm/DeploymentErrors";

import "brace/mode/yaml";
import "brace/theme/xcode";

interface IDeploymentFormProps {
  appCurrentVersion: string;
  appCurrentValues?: string;
  bindingsWithSecrets: IServiceBindingWithSecret[];
  chartName: string;
  kubeappsNamespace: string;
  namespace: string;
  releaseName: string;
  repo: string;
  error: Error | undefined;
  selected: IChartState["selected"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  fetchChartVersions: (id: string) => Promise<{}>;
  getBindings: (ns: string) => Promise<IServiceBindingWithSecret[]>;
  getChartVersion: (id: string, chartVersion: string) => Promise<void>;
  getChartValues: (id: string, chartVersion: string) => Promise<any>;
  clearRepo: () => any;
}

interface IDeploymentFormState {
  appValues?: string;
  isDeploying: boolean;
  valuesModified: boolean;
  version: string;
}

class UpgradeForm extends React.Component<IDeploymentFormProps, IDeploymentFormState> {
  public state: IDeploymentFormState = {
    appValues: this.props.appCurrentValues,
    isDeploying: false,
    valuesModified: false,
    version: this.props.appCurrentVersion,
  };

  public componentDidMount() {
    const {
      appCurrentVersion,
      chartName,
      fetchChartVersions,
      getBindings,
      getChartVersion,
      namespace,
      repo,
    } = this.props;
    const chartID = `${repo}/${chartName}`;
    fetchChartVersions(chartID);
    getChartVersion(chartID, appCurrentVersion);
    getBindings(namespace);
  }

  public componentDidUpdate(prevProps: IDeploymentFormProps) {
    const { selected, appCurrentVersion, appCurrentValues } = this.props;
    if (
      selected.version &&
      prevProps.selected.version &&
      selected.version !== prevProps.selected.version
    ) {
      // Version has changed
      if (selected.version.attributes.version === appCurrentVersion) {
        // The user has selected back the original version, use the current values
        if (!this.state.valuesModified) {
          // Only update the default values if the user has not modify them
          this.setState({ appValues: appCurrentValues });
        }
      }
    }
    if (selected.values && this.state.appValues && selected.values !== this.state.appValues) {
      // Values has been modified either because the user has edit them
      // or because the selected version is now different
      if (!this.state.valuesModified) {
        // Only update the default values if the user has not modify them
        if (this.state.version !== appCurrentVersion) {
          // Only use the default values if the version is not the original one
          this.setState({ appValues: selected.values });
        }
      }
    }
  }

  public render() {
    const { selected, bindingsWithSecrets, appCurrentVersion } = this.props;
    const { version, versions } = selected;
    const { appValues } = this.state;
    if (this.props.error) {
      return <DeploymentErrors {...this.props} version={appCurrentVersion} />;
    }
    if (!version || !versions || !versions.length || this.state.isDeploying) {
      return <div> Loading </div>;
    }
    return (
      <div>
        <form className="container padding-b-bigger" onSubmit={this.handleDeploy}>
          <div className="row">
            <div className="col-8">
              {this.props.error && (
                <DeploymentErrors {...this.props} version={version.attributes.version} />
              )}
            </div>
            <div className="col-12">
              <h2>
                {this.props.releaseName} ({this.props.chartName})
              </h2>
            </div>
            <div className="col-8">
              <div>
                <label htmlFor="chartVersion">Version</label>
                <select
                  id="chartVersion"
                  onChange={this.handleChartVersionChange}
                  value={version.attributes.version}
                  required={true}
                >
                  {versions.map(v => (
                    <option key={v.id} value={v.attributes.version}>
                      {v.attributes.version}{" "}
                      {v.attributes.version === this.props.appCurrentVersion ? "(current)" : ""}
                    </option>
                  ))}
                </select>
              </div>
              <div style={{ marginBottom: "1em" }}>
                <label htmlFor="values">Values (YAML)</label>
                <AceEditor
                  mode="yaml"
                  theme="xcode"
                  name="values"
                  width="100%"
                  onChange={this.handleValuesChange}
                  setOptions={{ showPrintMargin: false }}
                  editorProps={{ $blockScrolling: Infinity }}
                  value={appValues}
                />
              </div>
              <div>
                <button className="button button-primary" type="submit">
                  Submit
                </button>
                <button className="button" onClick={this.handleReselectChartRepo}>
                  Back
                </button>
              </div>
            </div>
            <div className="col-4">
              {bindingsWithSecrets.length > 0 && (
                <DeploymentBinding bindingsWithSecrets={bindingsWithSecrets} />
              )}
            </div>
          </div>
        </form>
      </div>
    );
  }

  public handleDeploy = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { releaseName, namespace, selected, upgradeApp, push } = this.props;
    this.setState({ isDeploying: true });
    const { appValues } = this.state;
    if (selected.version) {
      const deployed = await upgradeApp(selected.version, releaseName, namespace, appValues);
      this.setState({ isDeploying: false });
      if (deployed) {
        push(`/apps/ns/${namespace}/${releaseName}`);
      }
    }
  };

  public handleChartVersionChange = (e: React.FormEvent<HTMLSelectElement>) => {
    const { repo, chartName, getChartVersion, getChartValues } = this.props;
    const chartID = `${repo}/${chartName}`;
    this.setState({ version: e.currentTarget.value });
    getChartVersion(chartID, e.currentTarget.value);
    if (!this.state.valuesModified) {
      // Only update the default values if the user has not modify them
      getChartValues(chartID, e.currentTarget.value);
    }
  };

  public handleValuesChange = (value: string) => {
    this.setState({ appValues: value, valuesModified: true });
  };

  public handleReselectChartRepo = () => {
    this.props.clearRepo();
  };
}

export default UpgradeForm;
