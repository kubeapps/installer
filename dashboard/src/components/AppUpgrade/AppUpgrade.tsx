import * as React from "react";

import { RouterAction } from "connected-react-router";
import { hapi } from "../../shared/hapi/release";
import { IServiceBindingWithSecret } from "../../shared/ServiceBinding";
import { IAppRepository, IChartState, IChartVersion } from "../../shared/types";
import DeploymentErrors from "../DeploymentForm/DeploymentErrors";
import LoadingWrapper from "../LoadingWrapper";
import UpgradeForm from "../UpgradeForm";
import SelectRepoForm from "../UpgradeForm/SelectRepoForm";

interface IAppUpgradeProps {
  app: hapi.release.Release;
  bindingsWithSecrets: IServiceBindingWithSecret[];
  error: Error | undefined;
  repoError: Error | undefined;
  kubeappsNamespace: string;
  namespace: string;
  releaseName: string;
  version: string;
  repos: IAppRepository[];
  repo: IAppRepository;
  selected: IChartState["selected"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
  ) => Promise<boolean>;
  clearRepo: () => void;
  checkChart: (repo: string, chartName: string) => void;
  fetchChartVersions: (id: string) => Promise<IChartVersion[]>;
  getApp: (releaseName: string, namespace: string) => void;
  getBindings: (ns: string) => void;
  getChartVersion: (id: string, chartVersion: string) => void;
  getChartValues: (id: string, chartVersion: string) => void;
  push: (location: string) => RouterAction;
  fetchRepositories: () => void;
}

interface IAppUpgradeState {
  selectRepoForm: JSX.Element;
}

class AppUpgrade extends React.Component<IAppUpgradeProps, IAppUpgradeState> {
  public componentDidMount() {
    const { releaseName, getApp, namespace, fetchRepositories } = this.props;
    getApp(releaseName, namespace);
    fetchRepositories();
  }

  public render() {
    const { app, repos, error, repo } = this.props;
    if (
      !repos ||
      !app ||
      !app.chart ||
      !app.chart.metadata ||
      !app.chart.metadata.name ||
      !app.chart.metadata.version
    ) {
      if (!error) {
        return <LoadingWrapper />;
      } else {
        return (
          <DeploymentErrors
            {...this.props}
            chartName={(app.chart && app.chart.metadata && app.chart.metadata.name) || ""}
            repo={repo.metadata.name}
          />
        );
      }
    }
    if (!this.props.repo.metadata) {
      return (
        <div>
          <SelectRepoForm
            {...this.props}
            error={this.props.repoError}
            chartName={app.chart.metadata.name}
          />
        </div>
      );
    }
    return (
      <div>
        <UpgradeForm
          {...this.props}
          appCurrentVersion={app.chart.metadata.version}
          appCurrentValues={(app.config && app.config.raw) || ""}
          chartName={app.chart.metadata.name}
          repo={this.props.repo.metadata.name}
        />
      </div>
    );
  }
}

export default AppUpgrade;
