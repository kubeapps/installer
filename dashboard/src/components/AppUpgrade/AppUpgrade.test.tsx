import { shallow } from "enzyme";
import * as React from "react";

import { hapi } from "shared/hapi/release";
import { IAppRepository, IChartState } from "shared/types";
import itBehavesLike from "../../shared/specs";
import DeploymentErrors from "../DeploymentForm/DeploymentErrors";
import UpgradeForm from "../UpgradeForm";
import SelectRepoForm from "../UpgradeForm/SelectRepoForm";
import AppUpgrade from "./AppUpgrade";

const defaultProps = {
  app: {} as hapi.release.Release,
  bindingsWithSecrets: [],
  checkChart: jest.fn(),
  clearRepo: jest.fn(),
  error: undefined,
  fetchChartVersions: jest.fn(),
  fetchRepositories: jest.fn(),
  getApp: jest.fn(),
  getBindings: jest.fn(),
  getChartValues: jest.fn(),
  getChartVersion: jest.fn(),
  kubeappsNamespace: "kubeapps",
  namespace: "default",
  push: jest.fn(),
  releaseName: "foo",
  repo: {} as IAppRepository,
  repoError: undefined,
  repos: [],
  selected: {} as IChartState["selected"],
  upgradeApp: jest.fn(),
  version: "1.0.0",
};

itBehavesLike("aLoadingComponent", { component: AppUpgrade, props: defaultProps });

it("renders the repo selection form if not introduced", () => {
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
        } as hapi.release.Release
      }
      repos={[
        {
          metadata: { name: "stable" },
        } as IAppRepository,
      ]}
    />,
  );
  expect(wrapper.find(SelectRepoForm).exists()).toBe(true);
  expect(wrapper.find(DeploymentErrors).exists()).toBe(false);
  expect(wrapper.find(UpgradeForm).exists()).toBe(false);
  expect(wrapper).toMatchSnapshot();
});

it("renders an error if it exists", () => {
  const repo = {
    metadata: { name: "stable" },
  } as IAppRepository;
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      error={new Error("foo doesn't exists")}
      repos={[repo]}
      repo={repo}
    />,
  );
  expect(wrapper.find(DeploymentErrors).exists()).toBe(true);
  expect(wrapper.find(SelectRepoForm).exists()).toBe(false);
  expect(wrapper.find(UpgradeForm).exists()).toBe(false);
  expect(wrapper).toMatchSnapshot();
});

it("renders the upgrade form when the repo is available", () => {
  const repo = {
    metadata: { name: "stable" },
  } as IAppRepository;
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
        } as hapi.release.Release
      }
      repos={[repo]}
      repo={repo}
    />,
  );
  expect(wrapper.find(UpgradeForm).exists()).toBe(true);
  expect(wrapper.find(DeploymentErrors).exists()).toBe(false);
  expect(wrapper.find(SelectRepoForm).exists()).toBe(false);
  expect(wrapper).toMatchSnapshot();
});
