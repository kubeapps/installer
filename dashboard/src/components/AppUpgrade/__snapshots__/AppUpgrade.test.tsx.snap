// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
/>
`;

exports[`renders the repo selection form if not introduced 1`] = `
<SelectRepoForm
  chartName="bar"
  cluster="default"
  namespace="default"
/>
`;
